scripts/ci/pre_commit/pre_commit_update_versions.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import glob
+import json
+import os
+import re
+from os.path import abspath, dirname, join
+from re import Pattern
+
+import httpx
+
+AIRFLOW_SOURCES_DIR = abspath(join(dirname(__file__), os.pardir, os.pardir, os.pardir))
+
+
+def update_version(pattern: Pattern, version: str, file_path: str):
+    print(f"Replacing {pattern} to {version} in {file_path}")
+    with open(file_path, "r+") as f:
+        file_contents = f.read()
+        lines = file_contents.splitlines(keepends=True)
+        for i in range(0, len(lines)):
+            lines[i] = re.sub(pattern, fr'\g<1>{version}\g<2>', lines[i])
+        file_contents = "".join(lines)
+        f.seek(0)
+        f.truncate()
+        f.write(file_contents)
+
+
+REPLACEMENTS = {
+    r'(FROM apache/airflow:).*($)': "docs/docker-stack/docker-examples/extending/*/Dockerfile",
+    r'(apache/airflow:)[^-]*(\-)': "docs/docker-stack/entrypoint.rst",
+    r'(/constraints-)[^-]*(/constraints)': "docs/docker-stack/docker-examples/"
+    "restricted/restricted_environments.sh",
+    r'(AIRFLOW_VERSION=")[^"]*(" \\)': "docs/docker-stack/docker-examples/"
+    "restricted/restricted_environments.sh",
+}
+
+if __name__ == '__main__':
+    airflow_pypi_info = json.loads(httpx.get("https://pypi.org/pypi/apache-airflow/json").text)
```

Comment:
This is dangerous as it means that after a new version is released, all PRs have to do a rebase.

Reply:
That's right  any idea how we can improve it ?

Reply:
This will fail even before we release as for example when releasing `2.1.4`, this will use `2.1.3` instead as 2.1.4 won't be available until I push it.

I think this should match version in `setup.py` and just maintain a list of all the files that needs to be updated when the version is updated in setup.py. And if it contains "dev" in the version, it can then use `https://pypi.org/pypi/apache-airflow/json`

Reply:
Well. I deliberately used "latest-released" version.  Those dockerfiles actually **SHOULD** use 2.1.3 before we release 2.1.4 - because those docker-files are actually used  during tests and the dockerfiles are run through `docker build -f ..../Dockerfile`. 

If we start  using setup.py - that's the moment the tests will start to fail actually.

Reply:
We will only change to `2.1.4` when we will release `2.1.4` inside the feature branch; hence my comment about checking for "dev"

Reply:
We could also use "latest" (but then the dockerfile example is not best - because users should use a specfiic version).

It does not **really** matter if we use current or latest in this case - both should work fine because the tests are just testing if the "Dockerfiles" are valid and build fine, not whether particular version of Airflow works.

Actually the best way is to add "setup.py" version when the image is committed, but use "latest released" when we run tests. Which I might well do.

Reply:
Check above comment:

>And if it contains "dev" in the version, it can then use https://pypi.org/pypi/apache-airflow/json

Reply:
it is not just Dockerfile but also constraints, hence I will still suggest using my suggestion :)

Reply:
> That's right. Any idea how we can improve it ?

We can add a new step in the release process. We already have a similar one: https://github.com/apache/airflow/blob/main/dev/README_RELEASE_AIRFLOW.md#update-airflowconfig_templatesconfigyml-file

Reply:
Should be fixed now

* in the Dockerfiles we use the setup.py version
* but when we run tests, we dyanmically replace it with some released version (I fixed it to 2.1.3 - but we can irregularly update it too)

This way we test if the example dockerfiles are working fine (docker build succeds) even if the "source" version of airflow is not released yet.

Reply:
Small update. I changed it to use "latest released"  version for tests rather than fixed - this way we will not have to maintain it - right after releasing a new version to PyPI the `docekr builds` will start using the new version - but they should just continue building in people's PR. if they will start failing, that means that we should fix it (so that should solve the problem @mik-laj mentioned). 

This way we get an early warning system - if any of our "example extension Dockerfile" for any reason will stop working with latest released airflow, we will know it right after the new version has been released.

This is also nice as it will flag the cases in case we pushed a new release to PyPI but forgot to release the image. There is a slight race - if someone makes PR between PyPI and image release, the Dockerfile test might fail - but that is really rare and should not be a problem if the release manager has fast build-push cycle.

Reply:
Yep https://github.com/apache/airflow/pull/18335/checks?check_run_id=3636921660 - the tests work nicely building image based on "2.1.3" even if we have "2.2.0.dev0" in the Dockerfiles